api/services/syncing/syncing.go

```diff
@@ -49,18 +63,67 @@ type SyncConfig struct {
 	peers []*SyncPeerConfig
 }
 
-// GetStateSync returns the implementation of StateSyncInterface interface.
-func GetStateSync() *StateSync {
-	return &StateSync{}
+// CreateStateSync returns the implementation of StateSyncInterface interface.
+func CreateStateSync(ip string, port string) *StateSync {
+	stateSync := &StateSync{}
+	stateSync.selfip = ip
+	stateSync.selfport = port
+	stateSync.commonBlocks = make(map[int]*types.Block)
+	stateSync.lastMileBlocks = []*types.Block{}
+	return stateSync
 }
 
 // StateSync is the struct that implements StateSyncInterface.
 type StateSync struct {
+	selfip             string
+	selfport           string
 	peerNumber         int
 	activePeerNumber   int
-	blockHeight        int
+	commonBlocks       map[int]*types.Block
+	lastMileBlocks     []*types.Block // last mile blocks to catch up with the consensus
 	syncConfig         *SyncConfig
 	stateSyncTaskQueue *queue.Queue
+	syncMux            sync.Mutex
+}
+
+// AddLastMileBlock add the lastest a few block into queue for syncing
+func (ss *StateSync) AddLastMileBlock(block *types.Block) {
+	ss.lastMileBlocks = append(ss.lastMileBlocks, block)
+}
+
+// CloseConnections close grpc  connections for state sync clients
+func (ss *StateSync) CloseConnections() {
+	for _, pc := range ss.syncConfig.peers {
+		if pc.client != nil {
+			pc.client.Close()
+		}
+	}
+}
+
+// GetServicePort returns the service port from syncing port
+// TODO: really need use a unique ID instead of ip/port
+func GetServicePort(nodePort string) string {
+	if port, err := strconv.Atoi(nodePort); err == nil {
+		return fmt.Sprintf("%d", port+SyncingPortDifference)
+	}
+	os.Exit(1)
```

Comment:
Are you sure you want to exit here?  Please print out some error message if you exit here?

Reply:
This is original Minh's code, didn't touch it before.  I changed to just log it.